### ISSUE TITLE
Can't figure out how to represent unions

### ISSUE THREAD

**Issue by fabriziobertocci** [GaloisInc/daedalus#110]
The doc is very limited when explaining unions. Looking at the example doesn't help because I could not find a good simple case showing how to use a union. Newer version of daedalus appears to have a concept of `case` but there is no documentation.
I am trying to build a parser that can handle a stream of bytes describing a union as:
- 32-bit integer: discriminator value
- ... the serialized field
For example, given the following union (in IDL): 
```
  union Union switch(uint32) {
    case 1: octet  one;
    case 2: uint16 two;
    case 3: uint32 three;
};
```
The following stream of bytes:
```
01 00 00 00 : discriminator = 1, expect next field to be an octet
03          : value of field "one"
<padding>
Second sample - - - - - - - - - - 
03 00 00 00 : discriminator = 3, expect next field to be a 32-bit integer
05 00 00 00 : value of the field "three"
```
Moreover I would like the JSON output to be something like this:
[[ { "one": 1 }, { "three" :  5 } ]]

I have tried several way to describe it, but I was unable to get the result I wanted. EIther I would get some syntax error, internal error or the parser would produce an empty set. 
In all the cases, I have declared a top-level parser to deal with the multiple records:
```
def Union_Top = {
    ... -- dealing with the file header
    $$ = Many Union;
}
```

This case gives a parser error:
```
def Union = {
  @PadTo 4;
  @discr = LE32 as int;
  Choose {
     { discr is 1; one = UInt8; };
     { discr is 2; two = LE16;  };
     { discr is 3; three = LE32;  };
  }
}
```
This produces two empty objects [[ {}, {} ]]
```
def Union = {
  @PadTo 4;
  @discr = LE32 as int;
  Choose1 {
    { (discr == 1) is true; @one = UInt8; };
    { (discr == 2) is true; @two = LE16; };
    { (discr == 3) is true; @three = LE32; };
  }
}
```
Well... so this seems to succeed identifying the case, but produces no results because of the `@`. When I remove them I get an error `Type "Union.Union2" does not have field "two"`:
```
def Union = {
  @PadTo 4;
  @discr = LE32 as int;
  Choose1 {
    { (discr == 1) is true; one = UInt8; };
    { (discr == 2) is true; two = LE16; };
    { (discr == 3) is true; three = LE32; };
  }
}
```

This case produces an empty set [[ ]]:
```
def Case1 disc val= {
    (disc == val) is true;
    UInt8;
}
def Case2 disc val = {
    (disc == val) is true;
    LE16;
}
def Case3 disc val = {
    (disc == val) is true;
    LE32;
}
def Union = {
  @PadTo 4;
  @discr = LE32 as int;
  one = Case1 discr 0x01;
  two = Case2 discr 0x02;
  three = Case3 discr 0x03;
}
```

**Comment by yav** [comment]
Sorry I didn't see this until now.   The current way to this would be something like this:
```
def Union = {
  @PadTo 4;
  @discr = LE32 as int;
  Choose1 {
    one   = { (discr == 1) is true; UInt8; };
    two   = { (discr == 2) is true; LE16; };
    three = { (discr == 3) is true; LE32; };
  }
}
```
Basically the `Choose` and `Choose1` constructs have a "tagged" version that produces a union type.  I have been working on a change to Daedalus that would allow you to write such specifications more directly using `case` almost as in your first example, but that's not ready yet.

FYI, recent versions of Daedalus also support a different notation that uses "layout" (kind of like in Python) and allows for fewer braces.   Using that notation the example would look something like this:
```
def Union =
  block
    PadTo 4
    let discr = LE32
    Choose1
       one   = { (discr == 1) is true; UInt8 }
       two   = { (discr == 2) is true; LE16 }
       three = { (discr == 3) is true; LE32 }
```
Basically `block` is the same as the braces `{ .. }`.  Simlilarly, `Choose` and `Choose1` don't need the braces and semi-colons if you line up things properly.  Finally `let discr = LE32`  is the same as `@x = LE32`, just seems more readable.

Btw, I also removed the `as int` because it seems unneeded in this case.  In Daedalus, `int` is the type of arbitrary sized integers (think `Integer` in Java, or `gmp` integers in C/C++)---it is useful if you are concerned about overflow, but in this case we are just comparing things, so using the 32 bit number directly is simpler

**Comment by fabriziobertocci** [comment]
Thanks @yav  I tested both solution and they works... but the value I get back contains an extra `$`:
```
~/bin/daedalus --interp=cdr/Union.cdr --json Main.ddl
[[ { "$three": 1860308873 }
 , { "$three": 2908178047 }
 ]]
```
Where does that extra `$` come from?

Yes I saw the new syntax using indentation like Python (from the examples and test files). I wasn't ready to change my generated DDL to use the new structure because I could not find any documentation about that.

**Comment by yav** [comment]
That's part of the current schema for rendering semantic values into JSON---union tags have a `$` at the front to distinguish them from structs, which just use a label.    Every now and then you may see a `$$` too, that's for various built-in tags (if I recall correctly you'd see this if you make a value of type `Map`)

**Comment by fabriziobertocci** [comment]
Ah, so a `Choose1` constructs produces tags and not

**Comment by yav** [comment]
> Ah, so a `Choose1` constructs produces tags and not semantic value? Hmm ok I will adjust my JSON comparator to deal with that. Thanks for the help

The difference between `Choose1` (read "choose first") and `Choose` is that the former will pick the first matching branch (from the top), while the latter will consider all of them simultaneously.    In the cases where there is something to distinguish the branches they will produce the same result,  but `Choose1` is quite a bit more efficient, at least with the current implementation, so try to use that if possible.
